File: src/main/java/cspfj/generator/constraint/AbstractGenerator.java
package cspfj.generator.constraint;

import com.google.common.base.Function;
import com.google.common.base.Predicate;

import cspfj.constraint.Constraint;
import cspfj.exception.FailedGenerationException;
import cspfj.problem.BooleanDomain;
import cspfj.problem.Domain;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspom.variable.CSPOMVariable;

public abstract class AbstractGenerator implements Generator {

    public static Predicate<Variable> NULL_DOMAIN = new Predicate<Variable>() {
        @Override
        public boolean apply(Variable arg0) {
            return arg0.getDomain() == null;
        }
    };

    private final Problem problem;

    public Function<CSPOMVariable, Variable> cspomToCspfj = new Function<CSPOMVariable, Variable>() {
        @Override
        public Variable apply(CSPOMVariable input) {
            return problem.getVariable(input.getName());
        }
    };

    public AbstractGenerator(final Problem problem) {
        this.problem = problem;
    }

    public final Variable getSolverVariable(final CSPOMVariable variable) {
        return problem.getVariable(variable.getName());
    }

    public final void addConstraint(final Constraint constraint) {
        problem.addConstraint(constraint);
    }

    public final Variable addVariable(final String name, final Domain domain) {
        return problem.addVariable(name, domain);
    }

    public static void booleanDomain(final Variable variable)
            throws FailedGenerationException {
        if (variable.getDomain() == null) {
            variable.setDomain(new BooleanDomain());
        } else if (!(variable.getDomain() instanceof BooleanDomain)) {
            throw new FailedGenerationException(variable + " must be boolean");
        }
    }
}

File: src/main/scala/cspfj/generator/constraint/AbstractGenerator.scala
package cspfj.generator.constraint;

import cspfj.constraint.Constraint
import cspfj.generator.FailedGenerationException
import cspfj.problem.{ Variable, Problem, Domain, BooleanDomain }
import cspom.constraint.CSPOMConstraint
import cspom.variable.CSPOMVariable

abstract class AbstractGenerator(val problem: Problem) {
  def cspom2cspfj(variable: CSPOMVariable) = problem.variable(variable.name);

  def addConstraint(constraint: Constraint) = problem.addConstraint(constraint)

  def addVariable(name: String, domain: Domain) = problem.addVariable(name, domain)

  @throws(classOf[FailedGenerationException])
  def generate(constraint: CSPOMConstraint): Boolean
}

object AbstractGenerator {
  @throws(classOf[FailedGenerationException])
  def booleanDomain(variable: Variable) {
    if (variable.dom == null) {
      variable.dom = new BooleanDomain();
    } else if (!(variable.dom.isInstanceOf[BooleanDomain])) {
      throw new FailedGenerationException(variable + " must be boolean");
    }
  }

  /**
   * Sorts and remove duplicates from the given sequence to make eligible domain
   */
  def makeDomain(seq: Seq[Int]) =
    if (seq.isEmpty) { Nil }
    else {
      val itr = seq.sorted.reverseIterator
      itr.foldLeft(List(itr.next)) { (l, v) => if (l.head == v) l else v :: l }
    }

  def cartesianS[A, B, C](s0: Seq[A], s1: Seq[B], f: (A, B) => C) =
    for {
      i <- s0
      j <- s1
    } yield f(i, j)

  def cartesian[A](v0: Variable, v1: Variable, f: (Int, Int) => A) =
    cartesianS(v0.dom.allValues, v1.dom.allValues, f)

  def domainFrom(v0: Variable, v1: Variable, f: (Int, Int) => Int) =
    makeDomain(cartesian(v0, v1, f))
}